src/plone/restapi/services/auth/login.py

```diff
@@ -66,6 +66,10 @@ def reply(self):
                 )
             )
 
+        # Perform the same post-login actions as would happen when logging in through
+        # the Plone classic HTML login form.
+        self.context.unrestrictedTraverse("@@login")._post_login()
```

Comment:
I'm old, so checks out.  ;-)  I'm certainly open to suggestions.  I debated this myself when writing that line.  I opted for this because if a given deployment/application overrides that view I figured using any modified logic from the override would be less likely to be surprising.  Is there a more "modern" way to lookup a view in a context succinctly?  Is using the low-level ZCA adapter lookup considered  better?  (I'd find that hard to swallow personally.)  OTOH, the `._post_login()` method being used is a soft "private" method per Python conventions so I could appreciate an argument based on that except that what we're trying to accomplish here requires either calling that "private" method or violating DRY.

I'd love to hear any further thoughts.

Reply:
On the other hand, the title of the PR is "unify html form and API login". Can't have one without the other :)

Reply:
Although we opted for not using methods in browser views in most of the parts of plone.restapi it's true that we are use them anyways in some others. The rationale is to not to depend in them, specially if at some point in the future we get rid of those browser views (eg. moving out of the core the packages that include them, in an hypothetical Plone core "lite" with no Classic UI packages).

Reply:
I guess I'd feel more comfortable with `getMultiAdapter((self.context, self.request), name="login")`

Reply:
> I guess I'd feel more comfortable with `getMultiAdapter((self.context, self.request), name="login")`

I was never comfortable with this ZCA usage pattern as it doesn't seem explicit.  Couldn't there be other adapters registered for 2 objects that have the name `login` that aren't anything resembling a view?  I always felt that an adapter lookup should always specify the most specific provided interface that's appropriate for the use case.

That said, your suggestion *is* the lookup that ultimately happens in the `./plone.restapi/buildout.cfg` code.  Before that, however, `OFS.Traversable.Traversable.unrestrictedTraverse()` does a namespace lookup on the `@@` prefix which uses a `queryMultiAdapter((object, request), ITraversable, ns)` to get the actual code that actually performs the ultimate `queryMultiAdapter((self.context, self.request), name=name)` lookup to get the actual view.  So if a deployment/application has registered an alternate namespace traverser then the short-circuit straight to `getMultiAdapter((self.context, self.request), name="login")` in this PR would not use that alternate traverser which I think would be more surprising than the current implementation.

Having said all that, I don't actually care that much.  ;-)  Just repeat the request and I'll happily change to the suggested `queryMultiAdapter((self.context, self.request), name=name)`.

Reply:
There is no obvious solution to this IMHO, and all points raised here are valid claims. Using getMultiAdapter does sound like a more modern approach than unrestrictedTraverse. Depending on internal methods is a code smell indeed. Though, duplicating the code in the restapi code base also has its downsides (e.g. when the code changes in one of the sides). Bottom line is that if we remove browser views in the future, we will have to refactor things in restapi anyways I guess and in the meantime duplicating code has the potential to cause problems. Therefore I tend to agree with @rpatterson's tradeoff here and I'd just move to getMultiAdapter and add a few lines of explanation.